core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemUpgradeRequest.java

```diff
@@ -0,0 +1,139 @@
+ * limitations under the License.
+ */
+package com.b2international.snowowl.core.codesystem;
+
+import java.util.List;
+
+import javax.validation.constraints.NotNull;
+
+import org.apache.lucene.index.MergeTrigger;
+
+import com.b2international.commons.exceptions.BadRequestException;
+import com.b2international.snowowl.core.branch.Branch;
+import com.b2international.snowowl.core.domain.RepositoryContext;
+import com.b2international.snowowl.core.events.Request;
+import com.b2international.snowowl.core.identity.User;
+import com.b2international.snowowl.core.jobs.JobRequests;
+import com.b2international.snowowl.core.jobs.RemoteJobEntry;
+import com.b2international.snowowl.core.merge.Merge;
+import com.b2international.snowowl.core.repository.RepositoryRequests;
+import com.b2international.snowowl.core.uri.CodeSystemURI;
+import com.b2international.snowowl.core.uri.ResourceURIPathResolver;
+import com.b2international.snowowl.eventbus.IEventBus;
+
+/**
+ * @since 7.13.0 
+ */
+final class CodeSystemUpgradeRequest implements Request<RepositoryContext, String> {
+
+	private static final long serialVersionUID = 1L;
+
+	@NotNull
+	private final CodeSystemURI codeSystem;
+	
+	@NotNull
+	private final CodeSystemURI extensionOf;
+
+	private String codeSystemId;
+	
+	public CodeSystemUpgradeRequest(CodeSystemURI codeSystem, CodeSystemURI extensionOf) {
+		this.codeSystem = codeSystem;
+		this.extensionOf = extensionOf;
+	}
+	
+	void setCodeSystemId(String codeSystemId) {
+		this.codeSystemId = codeSystemId;
+	}
+	
+	@Override
+	public String execute(RepositoryContext context) {
+		if (!codeSystem.isHead()) {
+			throw new BadRequestException("Upgrades can not be started from CodeSystem versions.")
+				.withDeveloperMessage("Use '%s' only instead of '%s'", codeSystem.getCodeSystem(), codeSystem);
+		}
+
+		// get available upgrades 
+		final CodeSystem currentCodeSystem = CodeSystemRequests.prepareGetCodeSystem(codeSystem.getCodeSystem())
+				.setExpand(CodeSystem.Expand.AVAILABLE_UPGRADES + "()")
+				.build()
+				.execute(context);
+		final List<CodeSystemURI> availableUpgrades = currentCodeSystem.getAvailableUpgrades();
+		// report bad request if there are no upgrades available
+		if (availableUpgrades.isEmpty()) {
+			throw new BadRequestException("There are no upgrades available for the CodeSystem '%s'.", codeSystem.getCodeSystem());
+		}
+		
+		// or the selected extensionOf version is not present as a valid available upgrade
+		if (!availableUpgrades.contains(extensionOf)) {
+			throw new BadRequestException("Upgrades can only be performed to the next available version dependency.")
+				.withDeveloperMessage("Use '%s/<VERSION_ID>', where <VERSION_ID> is one of: '%s'", extensionOf.getCodeSystem(), availableUpgrades);
+		}
+		
+		// auto-generate the codeSystemId if not provided
+		// auto-generated upgrade IDs consist of the original CodeSystem's name and the new extensionOf dependency's URI
+		final String upgradeCodeSystemId;
```

Comment:
A check to see if the upgrade code system's short name is not already in use would be helpful, as a lot of work will be done before we find out if there was an issue with it.

Reply:
Branch creation will fail and will report a 409 issue in this case, which is enough for now, more complex checks will be implemented in the next 7.15.0 release.

Reply:
Isn't the original short name used for the new branch? That is different from the short name which will be created below.

Reply:
https://github.com/b2ihealthcare/snow-owl/blob/f2b19be624ddc9b9f26c45102085ea4b0e39a958/core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemUpgradeRequest.java#L99-L103